Automattic/jetpack
# Generate purchase token and append to all url's leading to checkout

This PR generates and adds a purchase token to outgoing urls leading to checkout. ie- "Upgrade" button CTA's and the "Plans" link. For use in the [Logged Out Visitor Checkout](pbtFFM-OD-p2) project.

**The idea and purpose behind this PR:**
When a Jetpack connected owner is logged into their wp-admin dashboard, a unique token is generated and saved to the wp_options table.  When a user (who is currently not logged in to WordPress.com) goes to upgrade/purchase a plan or product, this `purchasetoken` will be sent along to the Wordpress.com userless checkout page, where upon purchase, WordPress.com will request the site to validate the purchase token before completing the purchase.  After this validation step, the `purchasetoken` will be deleted from the options table, making it so a `purchasetoken` is only validated once before a new `purchacetoken` in generated again.  The purpose of this `purchasetoken` is so that the WordPress.com userless checkout(not logged into WordPress.com) can verify that the user originated from their Jetpack enabled site and they are the connection owner and the site is a valid jetpack connected site.



#### Changes proposed in this Pull Request:

* If `purchaseToken` doesn't exist in options table, one is created and then added to Redux `initialState` property.
* A `getPurchaseToken( state)` selector is added.
* A `purchasetoken` query param is added to all "Upgrade" button urls and to the "Plans" page url.
* ~~Added xmlrpc methods to retrieve & validate the client purchase token from Jetpack Server.~~ The endpoints were added in #19924.

#### Jetpack product discussion

* [Project Thread: Logged Out Visitor Checkout](pbtFFM-OD-p2).
* Asana Task: 1200152453830945-as-1200222681647748


#### Does this pull request change what data or activity we track or use?
No, I don't think so.



#### Testing instructions:






1. Checkout this PR and run locally or run this branch using the Jetpack Beta plugin.
2. Visit the Jetpack Dashboard page and verify that the "Plans" url now contains a `purchasetoken` param & value.
3. Scroll down further and verify that all the "Upgrade" button url's contain the `purchasetoken` param & value.
.

**Screenshot:** "Plans" url should now contain a `query` param with the `purchasetoken` key/val pair.
<img width="981" alt="Markup 2021-04-22 at 17 39 42" src="https://user-images.githubusercontent.com/11078128/115788811-24082b80-a392-11eb-868f-b9823012be2e.png">
.

**Screenshot:** "Upgrade" CTA's should now contain `purchasetoken` param.
<img width="985" alt="Markup 2021-04-22 at 17 34 47" src="https://user-images.githubusercontent.com/11078128/115788908-4dc15280-a392-11eb-80ec-01f1319f3438.png">
